### PR TITLE
Refactoring API

### PR DESCRIPTION
**Pull request by sebastienros**
- Ambient cache entries are automatically nested
- Removing `IEntryLink`
- Fixing nested cache entries invalidation
- Exposing factory helper methods

Requires @Tratcher, @lodejard

### CODE REVIEW

**Comment by sebastienros** [comment]
:up::date:

File: src/Microsoft.Extensions.Caching.Abstractions/ICacheEntry.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        /// Sets the value of the cache entry.
+        /// </summary>
+        /// <param name="value"></param>
+        ICacheEntry SetValue(object value);
+
+        /// <summary>
+        /// Sets the priority for keeping the cache entry in the cache during a memory pressure tokened cleanup.
+        /// </summary>
+        /// <param name="priority"></param>
+        ICacheEntry SetPriority(CacheItemPriority priority);
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/ICacheEntry.cs]: This SetFoo, return ICacheEntry style seems out of place on the base interface. We usually keep the interfaces simple and add sugar like this as extensions. I'll keep reviewing and see how it works out.

File: src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs [hunk 2]
```diff
@@ -21,34 +25,90 @@ public static TItem Get<TItem>(this IMemoryCache cache, object key)
...
 
-        public static object Set(this IMemoryCache cache, object key, object value, MemoryCacheEntryOptions options)
+        public static TItem Set<TItem>(this IMemoryCache cache, object key, TItem value, MemoryCacheEntryOptions options)
         {
-            return cache.Set(key, value, options);
+            using (var entry = cache.CreateEntry(key))
+            {
+                if (options != null)
+                {
+                    if (options.AbsoluteExpiration.HasValue)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs]: MemoryCacheEntryOptions and ICacheEntry need to be reconciled. The two should work mostly the same.

File: src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs [hunk 3]
```diff
@@ -21,34 +25,90 @@ public static TItem Get<TItem>(this IMemoryCache cache, object key)
...
+                }
+
+                entry.SetValue(value);
+            }
+
+            return value;
         }
 
-        public static TItem Set<TItem>(this IMemoryCache cache, object key, TItem value)
+        public static TItem GetOrCreate<TItem>(this IMemoryCache cache, object key, Func<ICacheEntry, TItem> factory)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs]: We removed factories like this over concerns that they were misleadingly similar to ConcurrentDictionary.GetOrCreate but didn't have it's locking. @Eilon?

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs [hunk 4]
```diff
@@ -1,47 +1,147 @@
...
+        internal IList<IChangeToken> _expirationTokens;
+        internal DateTimeOffset? _absoluteExpiration;
+        internal TimeSpan? _absoluteExpirationRelativeToNow;
+        internal TimeSpan? _slidingExpiration;
+        internal CacheItemPriority _priority;
 
         internal readonly object _lock = new object();
 
-        internal CacheEntry(
+        public CacheEntry(
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: internal

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs [hunk 5]
```diff
@@ -50,6 +150,23 @@ internal class CacheEntry
...
+                if (!_added)
+                {
+                    _notifyCacheEntryDisposed(this);
+                }
+            }
+            finally
+            {
+                CacheEntryHelper.LeaveScope();
+                PropageOptions(CacheEntryHelper.Current);
+                _added = true;
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: Set _added on line 159.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 6]
```diff
@@ -76,67 +76,69 @@ public object Set(object key, object value, MemoryCacheEntryOptions cacheEntryOp
...
 
             _entryLock.EnterWriteLock();
             try
             {
-                if (_entries.TryGetValue(key, out priorEntry))
+                if (_entries.TryGetValue(entry.Key, out priorEntry))
                 {
-                    _entries.Remove(key);
+                    _entries.Remove(entry.Key);
                     priorEntry.SetExpired(EvictionReason.Replaced);
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: Removing the existing entry should happen in SetEntry where you'll only need to take the write lock once.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 7]
```diff
@@ -302,6 +258,68 @@ private void EntryExpired(CacheEntry entry)
...
+            DateTimeOffset? absoluteExpiration = null;
+            if (entry._absoluteExpirationRelativeToNow.HasValue)
+            {
+                absoluteExpiration = utcNow + entry._absoluteExpirationRelativeToNow;
+            }
+            else if (entry._absoluteExpiration.HasValue)
+            {
+                if (entry._absoluteExpiration <= utcNow)
+                {
+                    throw new ArgumentOutOfRangeException(
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: SetEntry is only called from Dispose which should never throw. If you remove this I think CheckExpired will handle it as already expired.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 8]
```diff
@@ -456,24 +474,6 @@ private void ExpirePriorityBucket(int removalCountTarget, List<CacheEntry> expir
             }
         }
 
-        public void Dispose()
-        {
-            Dispose(true);
-        }
-
-        protected virtual void Dispose(bool disposing)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: Why were these Dispose methods removed? We do actually need them.

File: src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs (hunk 3 above)
Comment (Eilon) [inline src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs]: At this point, what are the differences between them that we want to ensure we avoid conflating?

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs (hunk 8 above)
Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: I have no idea! I must have let my machine unlocked too long ... I will add a unit test too.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs (hunk 7 above)
Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: We have a unit test that ensures an exception is thrown when adding an absolute expiration in the past, this case. Dispose actually throws in this case and fulfills the unit test as the SetEntry defines the utcNow to take into account, so I would preferably left it here. We can test the value when it's set as the current time is not valid at that moment.

Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: No, Dispose should never throw. Remove it and update the test to verify the value isn't added to the cache.

File: src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs]: This code should be simpler now with your updated ICacheEntry

File: src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs (hunk 3 above)
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs]: I recall the concern was around ConcurrentDictionary locking for the duration of the factory and only invoking it once, but the docs don't actually say that. https://msdn.microsoft.com/en-us/library/ee378677(v=vs.110).aspx

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs [hunk 9]
```diff
@@ -1,47 +1,167 @@
...
         internal CacheEntry(
             object key,
-            object value,
-            DateTimeOffset utcNow,
-            DateTimeOffset? absoluteExpiration,
-            MemoryCacheEntryOptions options,
+            Action<CacheEntry> notifyCacheEntryDisposed,
             Action<CacheEntry> notifyCacheOfExpiration)
         {
+            if (key == null)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: string.IsNullOrEmpty?

Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: Key is `object`

File: src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs (hunk 3 above)
Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs]: Should we solve it with documentation, on the interface itself, I know it has been an issue for Orchard as even myself was expecting some locking on the factory.

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs [hunk 10]
```diff
@@ -68,14 +205,14 @@ internal void SetExpired(EvictionReason reason)
 
         private bool CheckForExpiredTime(DateTimeOffset now)
         {
-            if (_absoluteExpiration.HasValue && _absoluteExpiration.Value <= now)
+            if (_absoluteExpiration <= now)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: What does `null <= now` return? false?

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 11]
```diff
@@ -37,7 +40,7 @@ public MemoryCache(IOptions<MemoryCacheOptions> optionsAccessor)
             var options = optionsAccessor.Value;
             _entries = new Dictionary<object, CacheEntry>();
             _entryLock = new ReaderWriterLockSlim();
-            _entryExpirationNotification = EntryExpired;
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: Why was this removed? I'm told we have to cache these explicitly because otherwise the compiler allocates them every call.

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs (hunk 10 above)
Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: Yes but I am reverting to prevent any confusion.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 12]
```diff
@@ -63,12 +66,8 @@ public int Count
...
         }
 
-        public IEntryLink CreateLinkingScope()
-        {
-            return EntryLinkHelpers.CreateLinkingScope();
-        }
-
-        public object Set(object key, object value, MemoryCacheEntryOptions cacheEntryOptions)
+        /// <inheritdoc />
+        public void Remove(object key)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: A lot of the diff in this view is just caused by method re-ordering, correct? Can you revert that where possible? E.g. Remove, Dispose, etc.?

**Comment by sebastienros** [comment]
:up: :date:

File: samples/MemoryCacheSample/Program.cs [hunk 13]
```diff
@@ -92,20 +92,20 @@ public static void Main()
             cts.Cancel();
 
             // Expire an entry if the dependent entry expires
-            using (var link = cache.CreateLinkingScope())
+            using (var entry = cache.CreateEntry("key1"))
             {
+                // expire this entry if the entry with key "key2" expires.
+                entry.SetValue("value1");
```
Comment (Tratcher) [inline samples/MemoryCacheSample/Program.cs]: Just show the property if you're not going to use the chaining.

File: src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs [hunk 14]
```diff
@@ -21,34 +25,78 @@ public static TItem Get<TItem>(this IMemoryCache cache, object key)
...
 
-        public static object Set(this IMemoryCache cache, object key, object value)
+        public static TItem Set<TItem>(this IMemoryCache cache, object key, TItem value)
         {
-            return cache.Set(key, value, new MemoryCacheEntryOptions());
+            return cache.Set(key, value, null);
         }
 
-        public static object Set(this IMemoryCache cache, object key, object value, MemoryCacheEntryOptions options)
+        public static TItem Set<TItem>(this IMemoryCache cache, object key, TItem value, MemoryCacheEntryOptions options)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/MemoryCacheExtensions.cs]: MemoryCacheEntryOptions is now a redundant temporary object. Would it make sense to have a few overloads of Set that just took the most common settings and bypassed creating the MemoryCacheEntryOptions? E.g.:
```
Set<TItem>(this IMemoryCache cache, object key, TItem value, DateTimeOffset absoluteExpiration)
Set<TItem>(this IMemoryCache cache, object key, TItem value, TimeSpan absoluteExpirationRelativeToNow)
Set<TItem>(this IMemoryCache cache, object key, TItem value, IChangeToken expirationToken)
```

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs [hunk 15]
```diff
@@ -1,47 +1,167 @@
...
+
         private static readonly Action<object> ExpirationCallback = ExpirationTokensExpired;
 
         private readonly Action<CacheEntry> _notifyCacheOfExpiration;
 
-        private readonly DateTimeOffset? _absoluteExpiration;
+        private readonly Action<CacheEntry> _notifyCacheEntryDisposed;
 
-        private IList<PostEvictionCallbackRegistration> _postEvictionCallbacks;
+        internal IList<PostEvictionCallbackRegistration> _postEvictionCallbacks;
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: private

**Comment by sebastienros** [comment]
:up: :date:

Added suggested simple extension methods.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 16]
```diff
@@ -302,6 +245,60 @@ private void EntryExpired(CacheEntry entry)
             StartScanForExpiredItems();
         }
 
+        private void SetEntry(CacheEntry entry)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: Move this up with CreateEntry.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 17]
```diff
@@ -63,109 +73,52 @@ public int Count
...
+                    _setEntry,
+                    _entryExpirationNotification
+                );
 
-            _entryLock.EnterWriteLock();
+            _entryLock.EnterReadLock();
             try
             {
-                if (_entries.TryGetValue(key, out priorEntry))
+                if (_entries.TryGetValue(entry.Key, out priorEntry))
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: Move into SetEntry. It's not replaced until ICacheEntry.Dispose is called.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 18]
```diff
@@ -174,30 +127,19 @@ public bool TryGetValue(object key, out object value)
...
-                            if (entry.Options.ExpirationTokens != null)
-                            {
-                                link.AddExpirationTokens(entry.Options.ExpirationTokens);
-                            }
-                            if (entry.Options.AbsoluteExpiration.HasValue)
-                            {
-                                link.SetAbsoluteExpiration(entry.Options.AbsoluteExpiration.Value);
-                            }
-                        }
+                        // When an entry is returned from an ambient entry,
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: This comment doesn't make sense.
When this entry is retrieved in the scope of creating another entry, that entry needs a copy of these expiration tokens.

**Comment by sebastienros** [comment]
:up: :date:

Moved more code in `SetEntry`, much better now.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 19]
```diff
@@ -63,77 +73,61 @@ public int Count
...
+            CacheEntry priorEntry;
 
             _entryLock.EnterWriteLock();
             try
             {
-                if (_entries.TryGetValue(key, out priorEntry))
+                
+                if (_entries.TryGetValue(entry.Key, out priorEntry))
                 {
-                    _entries.Remove(key);
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: Keep this remove. The new entry may already have expired and fail to overwrite, causing SetExpired to be fired again next time.

File: samples/MemoryCacheSample/Program.cs [hunk 20]
```diff
@@ -92,20 +92,20 @@ public static void Main()
...
-                    .AddEntryLink(link)
-                    .RegisterPostEvictionCallback(
+                entry.RegisterPostEvictionCallback(
                     (echoKey, value, reason, substate) =>
                     {
                         Console.WriteLine(echoKey + ": '" + value + "' was evicted due to " + reason);
-                    }));
+                    });
+
+                cache.Set("key2", "value2", new MemoryCacheEntryOptions()
```
Comment (Tratcher) [inline samples/MemoryCacheSample/Program.cs]: There's a new overload for this.

File: src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs [hunk 21]
```diff
@@ -314,7 +300,7 @@ private void StartScanForExpiredItems()
                     CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
             }
         }
-
+
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs]: nit: whitespace

File: test/Microsoft.Extensions.Caching.Memory.Tests/CacheServiceExtensionsTests.cs [hunk 22]
```diff
@@ -72,7 +72,7 @@ public void AddCaching_DoesNotReplaceUserRegisteredServices()
 
         private class TestMemoryCache : IMemoryCache
         {
-            public IEntryLink CreateLinkingScope()
+            public void Remove(object key)
```
Comment (Tratcher) [inline test/Microsoft.Extensions.Caching.Memory.Tests/CacheServiceExtensionsTests.cs]: Random reordering?

File: test/Microsoft.Extensions.Caching.Memory.Tests/MemoryCacheSetAndRemoveTests.cs [hunk 23]
```diff
@@ -131,6 +216,39 @@ public void SetOverwritesAndInvokesCallbacks()
         }
 
         [Fact]
+        public void SetOverwritesWithReplacedReason()
+        {
+            //System.Diagnostics.Debugger.Launch();
```
Comment (Tratcher) [inline test/Microsoft.Extensions.Caching.Memory.Tests/MemoryCacheSetAndRemoveTests.cs]: Remove

**Comment by sebastienros** [comment]
:up: :date:

**Comment by Tratcher** [comment]
:shipit:
